File: Src/Icm.ContextConsole/ContextTreeBuilder/ReflectionContextTreeBuilder.vb
Imports Icm.Tree
Imports Icm.Reflection

''' <summary>
''' Builds a two-level context tree by reflection, based on a root IContext.
''' </summary>
''' <remarks>
''' The first level is formed by the root context. The second level is formed by all the instantiable types (using Ninject)
''' of IContext, except for the type of the root context itself.)
''' </remarks>
Public Class ReflectionContextTreeBuilder
    Inherits TypeContextTreeBuilder

    Public Sub New(rootContextType As Type)
        MyBase.New()
        Dim typeRootNode = New TreeNode(Of Type)(rootContextType)
        Dim childContextTypes = GetAllImplementors(Of IContext)().Where(Function(ctxType) Not rootContextType.Equals(ctxType) AndAlso Not ctxType.IsNestedPrivate)
        typeRootNode.AddChildren(childContextTypes)
        Root = typeRootNode
    End Sub

End Class

File: Src/Icm.ContextConsole/ContextTreeBuilder/ReflectionContextTreeBuilder.cs
using System;
using System.Linq;
using Icm.Tree;
using static Icm.Reflection.ActivatorTools;

/// <summary>
/// Builds a two-level context tree by reflection, based on a root IContext.
/// </summary>
/// <remarks>
/// The first level is formed by the root context. The second level is formed by all the instantiable types (using Ninject)
/// of IContext, except for the type of the root context itself.) <see cref="Action(Of Integer)"></see>
/// </remarks>
public class ReflectionContextTreeBuilder : TypeContextTreeBuilder
{

	public ReflectionContextTreeBuilder(Type rootContextType) : base()
	{
		var typeRootNode = new TreeNode<Type>(rootContextType);
		var childContextTypes = GetAllImplementors<IContext>().Where(ctxType => !(rootContextType == ctxType) && !ctxType.IsNestedPrivate);
		typeRootNode.AddChildren(childContextTypes);
		Root = typeRootNode;
	}

}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
